lib/SILOptimizer/Mandatory/TFDeabstraction.cpp

```diff
@@ -1769,6 +1770,109 @@ static GraphOperationInst *tryToPromoteTensorFromScalars(
   return result;
 }
 
+/// This function transform the apply of the "__tf_tensor_from_scalar" function
+/// to a graph node (graph_op inst).
+///
+/// Similar to tryToPromoteTensorFromScalars(), the resulting graph node can be
+/// a "const", if its input scalar operand is a compile-time constant.
+/// Different from tryToPromoteTensorFromScalars(), we can also transform the
+/// apply into a "tfc.scalarToTensor" graph_op, which the partition pass then
+/// uses to do scalar promotion. See the comment above "func
+/// _TFTensorFromScalar<Scalar>" in Tensor.swift for more details.
+static GraphOperationInst *
+transformTensorFromScalar(ApplyInst *apply,
+                          const DenseMap<SILValue, SymbolicValue> &constants,
+                          GraphFunctionDeviceInfo &deviceInfo) {
+  assert(apply->getNumResults() == 1 &&
+         "tfc.scalarToTensor must have 1 result");
+  auto origResult = apply->getResults()[0];
+
+  assert(apply->getNumOperands() == 2);
+  auto scalarOperandAddr = apply->getOperand(1);
+  LLVM_DEBUG(llvm::dbgs() << "Converting __tf_tensor_from_scalar inst "
+                          << *apply
+                          << " with scalar operand : " << scalarOperandAddr
+                          << " of type " << scalarOperandAddr->getType()
+                          << " to a graph_op\n");
+  SILBuilder B(apply);
+  auto &ctx = B.getASTContext();
+
+  // First check to see if it's constant foldable.  If so, we can turn this into
+  // a Const node.
+  auto it = constants.find(scalarOperandAddr);
+  if (it != constants.end() && it->second.isConstant()) {
+    // Dig the element type out of the TensorHandle result type.
+    auto eltType =
+        getTensorHandleElementType(origResult->getType().getASTType());
+    // We use int32 as the element type of the zero-d shape array.
+    auto int32Ty =
+        ctx.getInt32Decl()->getDeclaredType()->getCanonicalType();
+    assert(it->second.getKind() == SymbolicValue::Address);
+
+    SmallVector<unsigned, 4> accessPath;
+    auto *scalarMemoryObject = it->second.getAddressValue(accessPath);
+    assert(accessPath.empty());
```

Comment:
I think it may be possible for there to be a non-empty access path, e.g.
```
func foo<T>(t: T) -> Tensor<T> {
  let tup = (t, t)
  return Tensor(tup.0)
}
```
The const evaluator might allocate an object for `tup`, and then return an address with access path into `tup` for the address of `tup.0`.

The PR I'm working on includes some helper functions for looking up values with access paths, so I'm fine if you submit it like this now and then someone updates this once the helpers exist.

Reply:
I tried this test case:
```swift
public func tuple(x: Float) {
  let tup = (x, x)
  _ = Tensor(tup.0)
}
```

And the deabstraction result is:
```
--- TFDeabstraction Result: $S3tmp5tuple1xySf_tF
// tuple(x:)
sil @$S3tmp5tuple1xySf_tF : $@convention(thin) (Float) -> () {
// %0                                             // users: %2, %2, %1
bb0(%0 : $Float):
  debug_value %0 : $Float, let, name "x", argno 1 // id: %1
  %2 = tuple (%0 : $Float, %0 : $Float)           // users: %4, %3
  debug_value %2 : $(Float, Float), let, name "tup" // id: %3
  %4 = tuple_extract %2 : $(Float, Float), 0      // user: %6
  %5 = alloc_stack $Float                         // users: %7, %17, %6, %19
  store %4 to %5 : $*Float                        // id: %6
  %7 = load %5 : $*Float                          // user: %9
  %8 = alloc_stack $Float                         // users: %10, %13, %9
  store %7 to %8 : $*Float                        // id: %9
  %10 = load %8 : $*Float                         // user: %11
  %11 = struct_extract %10 : $Float, #Float._value // user: %12
  %12 = graph_op "tfc.scalarToTensor"(%11 : $Builtin.FPIEEE32) {__device: "/job:localhost/replica:0/task:0/device:CPU:0"} : $TensorHandle<Float> // users: %14, %16, %20, %15, %18
  dealloc_stack %8 : $*Float                      // id: %13
  strong_retain %12 : $TensorHandle<Float>        // id: %14
  strong_release %12 : $TensorHandle<Float>       // id: %15
  strong_retain %12 : $TensorHandle<Float>        // id: %16
  destroy_addr %5 : $*Float                       // id: %17
  strong_release %12 : $TensorHandle<Float>       // id: %18
  dealloc_stack %5 : $*Float                      // id: %19
  strong_release %12 : $TensorHandle<Float>       // id: %20
  %21 = tuple ()                                  // user: %22
  return %21 : $()                                // id: %22
} // end sil function '$S3tmp5tuple1xySf_tF'
```

We can keep an eye on this issue.